cosmos/cache.py

```diff
@@ -400,3 +407,60 @@ def create_cache_profile(version: str, profile_content: str) -> Path:
     profile_yml_path = profile_yml_dir / DEFAULT_PROFILES_FILE_NAME
     profile_yml_path.write_text(profile_content)
     return profile_yml_path
+
+
+def is_cache_lockfile_enabled(project_dir: Path) -> bool:
+    project_lockfile = project_dir / PACKAGE_LOCK_YML
+    return project_lockfile.is_file()
+
+
+def _get_sha1_hash(yaml_file: Path) -> str:
+    """Read package-lock.yml file and return sha1_hash"""
+    with open(yaml_file) as file:
+        yaml_content = file.read()
+    data = yaml.safe_load(yaml_content)
+    sha1_hash: str = data.get("sha1_hash", "")
+    return sha1_hash
+
+
+def _get_latest_cached_lockfile(project_dir: Path) -> Path | None:
+    """
+    Retrieves the latest cached package-lock.yml for the specified project directory,
+    or creates and caches it if not already cached and hashes match.
+    """
+    cache_identifier = project_dir.name
+    project_lockfile = project_dir / PACKAGE_LOCK_YML
+    cached_package_lockfile = cache_dir / cache_identifier / PACKAGE_LOCK_YML
+
+    try:
+        if cached_package_lockfile.exists() and cached_package_lockfile.is_file():
+            project_sha1_hash = _get_sha1_hash(project_lockfile)
+            cached_sha1_hash = _get_sha1_hash(cached_package_lockfile)
+            if project_sha1_hash == cached_sha1_hash:
+                return cached_package_lockfile
+        else:
+            cached_lockfile_dir = cache_dir / cache_identifier
+            cached_lockfile_dir.mkdir(parents=True, exist_ok=True)
+            _safe_copy(project_lockfile, cached_package_lockfile)
+            return cached_package_lockfile
+    except OSError as e:
+        logger.warning(f"Error processing cached lockfile: {e}")
+    return None
+
+
+def _copy_cached_lockfile_to_project(cached_lockfile: Path, project_dir: Path) -> None:
+    """Copy the cached package-lock.yml to tmp project dir"""
+    lock_file = project_dir / PACKAGE_LOCK_YML
+    _safe_copy(cached_lockfile, lock_file)
+
+
+# TODO: Move this function at different location
+def _safe_copy(src: Path, dst: Path) -> None:
+    # Create a temporary file in the same directory as the destination
```

Comment:
@pankajastro could you please add a link to that article that discussed the issues with concurrency?

Reply:
Add docstring and also a reference. PTAL